compiler/rustc_passes/src/dead.rs

```diff
@@ -44,16 +44,52 @@ fn should_explore(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
     )
 }
 
-fn ty_ref_to_pub_struct(tcx: TyCtxt<'_>, ty: &hir::Ty<'_>) -> bool {
+fn struct_all_fields_are_public(tcx: TyCtxt<'_>, id: DefId) -> bool {
+    // treat PhantomData and positional ZST as public,
+    // we don't want to lint types which only have them,
+    // cause it's a common way to use such types to check things like well-formedness
+    tcx.adt_def(id).all_fields().all(|field| {
+        let field_type = tcx.type_of(field.did).instantiate_identity();
+        if field_type.is_phantom_data() {
+            return true;
+        }
+        let is_positional = field.name.as_str().starts_with(|c: char| c.is_ascii_digit());
+        if is_positional
+            && tcx
+                .layout_of(tcx.param_env(field.did).and(field_type))
+                .map_or(true, |layout| layout.is_zst())
+        {
+            return true;
+        }
+        field.vis.is_public()
+    })
+}
+
+/// check struct and its fields are public or not,
+/// for enum and union, just check they are public,
+/// and doesn't solve types like &T for now, just skip them
+fn ty_ref_to_pub_struct(
+    tcx: TyCtxt<'_>,
+    ty: &hir::Ty<'_>,
+) -> (bool, /* public and all fields are public */ bool) {
```

Comment:
I think the jump from anonymous bool to a multiple-boolean thing justifies naming the fields, rather than relying on a comment embedded in the return type here.

I.e. please replace the `(bool, bool)` here with something like

```rust
struct Publicness { ty_is_public: bool, ty_and_all_fields_are_public: bool }
```

(because that will help the calling code be more self-documenting, rather than forcing a human  to interpret the meaning of `!ty_ref_to_pub_struct(self.tcx, impl_ref.self_ty).1`)